Apromore-Core-Components/Apromore-Portal/src/main/java/org/apromore/portal/ApromoreZKListener.java

```diff
@@ -0,0 +1,200 @@
+            return;
+        }
+
+        final HttpServletRequest httpServletRequest = (HttpServletRequest) exec.getNativeRequest();
+        final HttpServletResponse httpServletResponse = (HttpServletResponse) exec.getNativeResponse();
+
+        // If we are not using Keycloak, we don't have to manage JWTs
+        final ConfigBean config = (ConfigBean) SpringUtil.getBean("portalConfig");
+        final boolean usingKeycloak = config.isUseKeycloakSso();
+        if (!usingKeycloak) {
+            LOGGER.debug("Skipping JWT check because not using Keycloak");
+
+            return;
+        }
+
+        final String appAuthHeader = JwtHelper.readCookieValue(httpServletRequest, JWT_COOKIE_NAME);
+        LOGGER.trace("Read App_Auth cookie: {}", appAuthHeader);
+        final String signedAuthHeader = JwtHelper.readCookieValue(httpServletRequest, JwtHelper.SIGNED_AUTH_HEADER_KEY);
+
+        try {
+            final JWTClaimsSet jwtClaimsSet = JwtHelper.getClaimsSetFromJWT(appAuthHeader);
+
+            final String issuedAtStr = jwtClaimsSet.getStringClaim(JwtHelper.STR_JWT_KEY_ISSUED_AT);
+            LOGGER.debug("issuedAtStr {}", issuedAtStr);
+            final Date issuedAtDate = jwtClaimsSet.getIssueTime();
+            LOGGER.trace("issuedAtDate {}", issuedAtDate);
+            final Date expiryAtDate = jwtClaimsSet.getExpirationTime();
+            LOGGER.debug("expiryAtDate {}", expiryAtDate);
+
+            if (!checkDigitalSignature(appAuthHeader, signedAuthHeader)) {
+                throw new Exception("JWT was not correctly signed: appAuthHeader " + appAuthHeader +
+                    " signedAuthHeader " + signedAuthHeader);
+
+            } else if (JwtHelper.isJwtExpired(jwtClaimsSet, issuedAtDate, expiryAtDate)) {
+                LOGGER.info("JWT is expired, signing this session out");
+                signOut(exec.getSession());
+
+            } else if (JwtHelper.doesJwtExpiryWithinNMinutes(jwtClaimsSet, config.getMinutesUntilExpiryBeforeSessionRefresh())) {
+                // If the token is close to expiry, refresh it
+                LOGGER.info("JWT is close to expiry, refreshing it");
+                refreshSessionTimeout(jwtClaimsSet, exec, appAuthHeader, httpServletResponse);
+
+            } else {
+                LOGGER.debug("JWT is fresh");
+            }
+        } catch (final Exception e) {
+            LOGGER.error("JWT security token check failed; terminating session", e);
+            signOut(exec.getSession());
+        }
+    }
+
+    /**
+     * @param jwtStr  a plaintext JSON web token
+     * @param base64EncodedAndSignedJwtStr  an encoded and signed JSON web token
+     * @return whether  <var>base64EncodedAndSignedJwtStr</var> has content matching
+     *     <var>jwtStr</var> and is correctly encoded and signed by the key pair
+     *     identified by {@link KEY_ALIAS}
+     */
+    private boolean checkDigitalSignature(final String jwtStr, final String base64EncodedAndSignedJwtStr) {
```

Comment:
public static boolean isSignedStrVerifiable()
y is this method in jwthelper not used.

Reply:
Changed to use JwtHelper method and better named method.